### PR TITLE
Improve safety when deleting files from the store

### PR DESCRIPTION
**Pull request by s1monw**
Today if we delete files from the index directory we never acquire the
write lock. Yet, for safety reasons we should get the write lock before
we modify / delete any files. Where we can we should leave the deletion
to the index writer and only delete that are necessary to delete ourself.

### CODE REVIEW

**Comment by s1monw** [comment]
I updated this to fix other places as well where we need the write lock and added more safety. @mikemccand @rmuir can you take a look please

**Comment by rmuir** [comment]
+1

File: src/main/java/org/elasticsearch/index/store/Store.java [hunk 1]
```diff
@@ -558,17 +542,28 @@ private static final void failIfCorrupted(Directory directory, ShardId shardId)
...
-                        dir.deleteFile(reason, existingFile);
-                    } catch (Exception ex) {
-                        logger.debug("failed to delete file [{}]", ex, existingFile);
-                        // ignore, we don't really care, will get deleted later on
+                if (existingFile.equals(IndexWriter.WRITE_LOCK_NAME) || Store.isChecksum(existingFile) || sourceMetaData.contains(existingFile)) {
+                    continue; // don't delete snapshot file, or the checksums file (note, this is extra protection since the Store won't delete checksum)
+                }
+                try {
+                    dir.deleteFile(reason, existingFile);
+                } catch (FileNotFoundException | NoSuchFileException ex) {
```
Comment (mikemccand) [inline src/main/java/org/elasticsearch/index/store/Store.java]: Hmm if we have the write lock on this dir, why would we hit FNFE/NSFE?

File: src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 2]
```diff
@@ -152,25 +150,30 @@ private static SegmentInfos readSegmentInfos(String segmentsFileName, Directory
...
+                    }
                 }
             }
-        }
-        assert SegmentInfos.getLastCommitSegmentsFileName(directory).equals(segmentsFileName);
-        if (foundSegmentFiles == 0) {
-            throw new IllegalStateException("no commit found in the directory");
+            assert SegmentInfos.getLastCommitSegmentsFileName(directory).equals(segmentsFileName);
+            if (foundSegmentFiles == 0) {
+                throw new IllegalStateException("no commit found in the directory");
```
Comment (mikemccand) [inline src/main/java/org/elasticsearch/common/lucene/Lucene.java]: Maybe add a comment in javadoc saying index must have a commit?

**Comment by mikemccand** [comment]
Left a couple comments, otherwise LGTM!

File: src/main/java/org/elasticsearch/index/store/Store.java (hunk 1 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/index/store/Store.java]: good question - I will remove

File: src/main/java/org/elasticsearch/common/lucene/Lucene.java (hunk 2 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/common/lucene/Lucene.java]: will do

**Pull request merged**